src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
 
+using System.Diagnostics;
+
 namespace System.Collections.Generic
 {
     /// <summary>Internal helper functions for working with enumerables.</summary>
     internal static class EnumerableHelpers
     {
-        /// <summary>Converts an enumerable to an array using the same logic as does List{T}.</summary>
+        /// <summary>Converts an enumerable to an array.</summary>
         /// <param name="source">The enumerable to convert.</param>
         /// <returns>The resulting array.</returns>
         internal static T[] ToArray<T>(IEnumerable<T> source)
         {
-            int count;
-            T[] results = ToArray(source, out count);
-            Array.Resize(ref results, count);
-            return results;
+            Debug.Assert(source != null);
+
+            var collection = source as ICollection<T>;
+            if (collection != null)
+            {
+                int count = collection.Count;
+                if (count == 0)
+                    return Array.Empty<T>();
+                
+                var result = new T[count];
+                collection.CopyTo(result, arrayIndex: 0);
+                return result;
+            }
+
+            // Generic methods are compiled per-instantiation for value types.
+            // Do not force the jit to compile a bunch of code it will never
+            // use if we only go down the ICollection path.
+            return LazyToArray(source);
+        }
+
+        private static T[] LazyToArray<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(!(source is ICollection<T>), $"We should have gone down the optimized route if {nameof(source)} was a regular collection.");
+
+            const int InitialCapacity = 4;
+            const int BufferListThreshold = 32;
+
+            using (var en = source.GetEnumerator())
+            {
+                if (!en.MoveNext())
+                    return Array.Empty<T>();
+
+                T[] buffer = null;
+                int index = 0; // index into the buffer we are reading into
+
+                do
+                {
+                    if (buffer == null)
+                    {
+                        // First iteration: allocate an array with the initial capacity
+                        buffer = new T[InitialCapacity];
+                    }
+                    else
+                    {
+                        // Resize from a previous iteration
+                        T[] last = buffer;
+                        int newLength = last.Length * 2;
+                        buffer = new T[newLength];
+
+                        // We are using a for-loop instead of Array.Resize under the assumption
+                        // we are copying < 32 elements.
+                        Debug.Assert(last.Length < BufferListThreshold);
+                        for (int i = 0; i < last.Length; i++) buffer[i] = last[i];
```

Comment:
We should just use Array.Resize / Array.Copy and let them employ appropriate optimizations.